settings_org: Add setting for invite link generation policy

dmenz opened the pull request:
Add setting to allow controlling who can create multi-use invite links

Fixes: #15159

The changes were mostly inspired by how the invite by e-mail setting is already done. 


**Screenshots and screen captures:**

Before:
![image](https://user-images.githubusercontent.com/52861789/207203938-f9bae73b-49ed-4179-8301-7dfc4fa4e0fb.png)

After:
![image](https://user-images.githubusercontent.com/52861789/207204361-a12bb555-1e8d-48aa-994a-db484d132660.png)

Help center section "Change who can send invitations" before:
![image](https://user-images.githubusercontent.com/52861789/217601188-7fb22e29-5bab-41b0-bd40-c82c77eb418b.png)

Help center section "Change who can send invitations" after:
![image](https://user-images.githubusercontent.com/52861789/217597933-b34f39e7-c3c3-4cfd-8f90-13dc84a01884.png)

Help center section "Create an invitation link" before:

![image](https://user-images.githubusercontent.com/52861789/217601591-cfbd19b2-b661-4f8e-a68c-cd2d3cff2d32.png)

Help center section "Create an invitation link" after:
![image](https://user-images.githubusercontent.com/52861789/217601988-dea57e40-f23d-4cf8-98fe-de378a283246.png)




<details>
<summary>Manual Tests Realized</summary>

1. Only owner can generate invite links when the policy is set to Nobody.

2. Members can generate invite links when the policy is set to Admins, Moderators and Members.

3. Members can't generate invite links when the policy is set to Admins. 
</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

dmenz commented:
The puppeteer tests are all passing on my local environment. I think the failure on the CI isn't related to the changes of this PR.

alya commented:
Thanks! Could you please add screenshots of the help center changes as well?

alya commented:
You will also need to remove the merge commit; see [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for commit discipline guidelines.

alya commented:
Hm, I'm not sure why I'm running into a merge conflict when I run `tools/fetch-rebase-pull-request 23849`.

dmenz commented:
@alya We removed the merge commit, added screenshots of the help center changes and checked fetch-rebase-pull-request is working without merge conflicts. Could you check it again, please?

alya commented on help/restrict-account-creation.md:
```diff
@@ -53,15 +53,14 @@ Regardless of whether invitations are required, you can:
-while permission to send out individual email invitations is configurable, creating
-*reusable* invitation links is always limited to administrators.
+permission to send out individual email invitations and create invitation links are configurable.
 
```
Let's make that "permissions".

dmenz replied:
Done

alya commented on help/restrict-account-creation.md:
```diff
@@ -53,15 +53,14 @@ Regardless of whether invitations are required, you can:
 1. Under **Joining the organization**, configure
-   **Who can invite users to this organization**.
+   **Who can send e-mail invitations to new users** or **Who can create invite links**.
 
```
Let's make that "and" rather than "or".

dmenz replied:
Done

alya commented:
Hmm, I tried to test again, but for some reason provisioning failed.

dmenz commented:
Ready for another review!

alya commented:
Provisioning still fails for me, so I can't test this PR:

```
New digest written to: /srv/zulip/var/eb9a49c3-318e-45f8-a160-eafd183b0ce8/migrations_hash_zulip
+ env DJANGO_SETTINGS_MODULE=zproject.settings ZULIP_DB_NAME=zulip ./manage.py migrate --no-input
Conflicting migrations detected; multiple leaf nodes in the migration graph: (0426_realm_create_multiuse_invite_to_realm_policy, 0430_fix_audit_log_objects_for_group_based_stream_settings in zerver).
To fix them run 'python manage.py makemigrations --merge'

Error running a subcommand of /srv/zulip/tools/lib/provision_inner.py: env DJANGO_SETTINGS_MODULE=zproject.settings ZULIP_DB_NAME=zulip ./manage.py migrate --no-input
Actual error output for the subcommand is just above this.


Provisioning failed (exit code 1)!

* Look at the traceback(s) above to find more about the errors.
* Resolve the errors or get help on chat.
* If you can fix this yourself, you can re-run tools/provision at any time.
* Logs are here: zulip/var/log/provision.log
```

timabbott commented:
`tools/renumber-migrations` is the way to address that. @dmenz can you rebase this including running that? See https://zulip.readthedocs.io/en/latest/subsystems/schema-migrations.html for details.

dmenz commented:
I've rebased the branch and renumbered the migration accordingly. Could you check it again, please?

alya commented:
Thanks, this seems like a good start! Playing with it, we need to improve the UX for the user invite modal if the user has partial permissions (only email or link invites). In this situation, we should disable the email input field or the "Generate invite link" link as appropriate, and show a tooltip explaining what's going on:

- You do not have permissions to send email invitations in this organization.
- You do not have permissions to generate invite links in this organization.

alya commented:
And when the user doesn't have either invite permission, we should hide the "Invite more users" links, just as we do now.

dmenz commented:
Thank you for the review! 
I'm thinking what is a good way to visually indicate that the link generation is disabled.

Currently the "Generate invite link" button is styled as a link and it becomes a marked checkbox when it's clicked:

![image](https://user-images.githubusercontent.com/52861789/222309236-2948370b-b2cc-4086-a145-1bd649de002e.png)
![image](https://user-images.githubusercontent.com/52861789/222309456-c4239eb5-876b-4785-98b3-f41088d790fa.png)

And when the user doesn't have permissions to generate invite links this button just doesn't show up.


One way I think to represent that this option is disabled is to make it always visible and always a checkbox, so we can disable the checkbox when the user doesn't have permissions, which makes it gray and unmarkable like this:

![image](https://user-images.githubusercontent.com/52861789/222315774-946256a2-bf4f-4867-90d3-a351cbbfc1a5.png)
![image](https://user-images.githubusercontent.com/52861789/222316732-b575803c-26ea-4f87-8073-be9e50dabe2e.png)


What are your thoughts on this, @alya ?

alya commented:
Hm, I don't have a clear answer here. Could you please start a discussion in the the #design stream in the development community. You can post your ideas from the comment above as a starting point.

dmenz commented:
Sure! [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Invite.20link.20generation.20checkbox).

dmenz commented:
Now that  #24692 is finished, can we get back to this?

sahil839 commented:
@dmenz Thanks for all your work and efforts on this. We appreciate it, but we have decided to make this a group permission setting in #25884 and the work for it is very close to being merged.

I am closing this PR for now and thanks again!
